File: 1807. Evaluate the Bracket Pairs of a String.js
// 1807. Evaluate the Bracket Pairs of a String

/**
 * @param {string} s
 * @param {string[][]} knowledge
 * @return {string}
 */
var evaluate = function(s, knowledge) {
  const knowMap = {}
  for(let i = 0; i < knowledge.length; ++i) {
      knowMap[knowledge[i][0]] = knowledge[i][1]
  }
  return s.replaceAll(/\((\w+)\)/g, ($0, $1) => {
      // console.log($0, $1)
      return knowMap[$1] ? knowMap[$1] : '?'
  })
};

File: 1807. Evaluate the Bracket Pairs of a String.py
# 1807. Evaluate the Bracket Pairs of a String

from typing import List
import re

class Solution:
    def evaluate(self, s: str, knowledge: List[List[str]]) -> str:
        knowMap = {}

        for [key, value] in knowledge:
            knowMap[key] = value

        def repl(m):
            key = m.group(1)
            return knowMap[key] if key in knowMap else '?'

        return re.sub(r'\((\w+)\)', repl, s)

solution = Solution()
solution.evaluate("(name)is(age)yearsold", [["name","bob"],["age","two"]])
